File: backend/app/modules/sales/controller.py
from flask import Blueprint, jsonify, request

from app.modules.sales.service import SaleService
from app.routes.auth import login_required

sales_bp = Blueprint("sales", __name__, url_prefix="/api/sales")
service = SaleService()


@sales_bp.get("")
@login_required()
def list_sales(current_user):
    return jsonify({"items": service.list_sales(request.args.to_dict())}), 200


@sales_bp.post("")
@login_required()
def create_sale(current_user):
    return jsonify(service.create_sale(request.get_json(force=True), current_user["id"])), 201


@sales_bp.patch("/<int:sale_id>")
@login_required(required_profiles={"ADMIN"})
def update_sale(current_user, sale_id: int):
    return jsonify(service.update_sale(sale_id, request.get_json(force=True))), 200


@sales_bp.delete("/<int:sale_id>")
@login_required(required_profiles={"ADMIN"})
def delete_sale(current_user, sale_id: int):
    service.delete_sale(sale_id)
    return jsonify({"message": "Venda excluida com sucesso."}), 200

File: backend/src/modules/sales/controller.ts
import type { Express } from "express";

import { asyncHandler, getQueryParams } from "../../core/http";
import { loginRequired } from "../../routes/auth";
import type { SerializedUser } from "../../types";
import { SaleService } from "./service";

const service = new SaleService();

export function registerSalesRoutes(app: Express) {
  app.get(
    "/api/sales",
    loginRequired(),
    asyncHandler(async (req, res) => {
      res.status(200).json({ items: await service.listSales(getQueryParams(req.query)) });
    }),
  );

  app.post(
    "/api/sales",
    loginRequired(),
    asyncHandler(async (req, res) => {
      const currentUser = res.locals.currentUser as SerializedUser;
      res.status(201).json(await service.createSale(req.body ?? {}, currentUser.id));
    }),
  );

  app.patch(
    "/api/sales/:saleId",
    loginRequired(new Set(["ADMIN"])),
    asyncHandler(async (req, res) => {
      res.status(200).json(await service.updateSale(Number(req.params.saleId), req.body ?? {}));
    }),
  );

  app.delete(
    "/api/sales/:saleId",
    loginRequired(new Set(["ADMIN"])),
    asyncHandler(async (req, res) => {
      await service.deleteSale(Number(req.params.saleId));
      res.status(200).json({ message: "Venda excluida com sucesso." });
    }),
  );
}
